### ISSUE TITLE
[train] Big performance hit when TensorFlow trainer is not scheduled on head node

### ISSUE THREAD

**Issue by krfricke** [ray-project/ray#32509]
### What happened + What you expected to happen

I'm running the Quick start example script on a cluster with 3 nodes a 8 CPUs.

The training time can be vastly different:

```
2023-02-13 18:17:26,474 INFO tune.py:762 -- Total run time: 34.41 seconds (33.59 seconds for the tuning loop).
```

vs.

```
2023-02-13 18:18:58,509 INFO tune.py:763 -- Total run time: 120.22 seconds (119.19 seconds for the tuning loop).
```

It seems like this depends mostly on where the trainable actor is scheduled. If it's scheduled on the head node, it runs fast. If it is scheduled on a worker node, it's slow.

When it's slow, I see a lot of these:

```
2023-02-13 18:18:44,975 WARNING util.py:244 -- The `process_trial_save` operation took 1.714 s, which may be a performance bottleneck.`
```

presumably because checkpoints are synced to the head node.

Ideally we would co-locate the trainable with the driver, especially when not using cloud storage.

### Versions / Dependencies

2.2.0

### Reproduction script

Full script:

```
import ray

# Load data.
dataset = ray.data.read_csv("s3://anonymous@air-example-data/breast_cancer.csv")

# Split data into train and validation.
train_dataset, valid_dataset = dataset.train_test_split(test_size=0.3)

# Create a test dataset by dropping the target column.
test_dataset = valid_dataset.drop_columns(cols=["target"])

import numpy as np

from ray.data.preprocessors import Concatenator, Chain, StandardScaler

# Create a preprocessor to scale some columns and concatenate the result.
preprocessor = Chain(
    StandardScaler(columns=["mean radius", "mean texture"]),
    Concatenator(exclude=["target"], dtype=np.float32),
)

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

from ray.air import session
from ray.air.config import ScalingConfig
from ray.air.integrations.keras import Callback as KerasCallback
from ray.train.tensorflow import TensorflowTrainer


def create_keras_model(input_features):
    return keras.Sequential(
        [
            keras.Input(shape=(input_features,)),
            layers.Dense(16, activation="relu"),
            layers.Dense(16, activation="relu"),
            layers.Dense(1),
        ]
    )


def train_loop_per_worker(config):
    batch_size = config["batch_size"]
    lr = config["lr"]
    epochs = config["num_epochs"]
    num_features = config["num_features"]

    # Get the Ray Dataset shard for this data parallel worker,
    # and convert it to a Tensorflow Dataset.
    train_data = session.get_dataset_shard("train")

    strategy = tf.distribute.MultiWorkerMirroredStrategy()
    with strategy.scope():
        # Model building/compiling need to be within `strategy.scope()`.
        multi_worker_model = create_keras_model(num_features)
        multi_worker_model.compile(
            optimizer=tf.keras.optimizers.SGD(learning_rate=lr),
            loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
            metrics=[
                tf.keras.metrics.BinaryCrossentropy(
                    name="loss",
                )
            ],
        )

    for _ in range(epochs):
        tf_dataset = train_data.to_tf(
            feature_columns="concat_out", label_columns="target", batch_size=batch_size
        )
        multi_worker_model.fit(
            tf_dataset,
            callbacks=[KerasCallback()],
            verbose=0,
        )


num_features = len(train_dataset.schema().names) - 1

trainer = TensorflowTrainer(
    train_loop_per_worker=train_loop_per_worker,
    train_loop_config={
        "batch_size": 128,
        "num_epochs": 50,
        "num_features": num_features,
        "lr": 0.0001,
    },
    scaling_config=ScalingConfig(
        num_workers=2,  # Number of data parallel training workers
        use_gpu=False,
        trainer_resources={"CPU": 6},  # so that the example works on Colab.
        resources_per_worker={"CPU": 8},
    ),
    datasets={"train": train_dataset},
    preprocessor=preprocessor,
)

result = trainer.fit()
print(f"Last result: {result.metrics}")
# Last result: {'loss': 8.997025489807129, ...}
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by cszhu** [comment]
This P2 issue has seen no activity in the past 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please comment and remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
